server/src/core/distribution-report/distribution-report-service.ts

```diff
@@ -0,0 +1,98 @@
+import { Db, Collection } from 'mongodb';
+import { DistributionReportService, DistributionReportArgs, SmsAndEmailMessages } from './types';
+import { rethrowIfAppError, createDbOpFailedError } from '../error';
+import { User } from '../user';
+import { extractFirstName } from '../util';
+import { DistributionReport } from '../payment';
+
+const COLLECTION = 'distribution_reports';
+
+export class DistributionReports implements DistributionReportService {
+  private db: Db;
+  private collection: Collection<DistributionReport>;
+  private args: DistributionReportArgs;
+
+  constructor(db: Db, args: DistributionReportArgs) {
+    this.db = db;
+    this.collection = db.collection(COLLECTION);
+    this.args = args;
+  }
+
+  async sendDistributionReportsToDonors(): Promise<void> {
+    try {
+      const reports: DistributionReport[] = await this.args.transactions.generateDistributionReportDocs();
+      await this.sendDistributionReportMessages(reports);
+      if (reports.length) {
+        await this.collection.insertMany(reports);
+      }
+    }
+    catch (e) {
+      console.error("Error occured: ", e.message);
+      rethrowIfAppError(e);
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  private async sendDistributionReportMessages(reportDocs: DistributionReport[]): Promise<void> {
+    try {
+      reportDocs.forEach(async (report: DistributionReport) => {
+        const donor = await this.args.users.getById(report.donor);
+        const beneficiaries = await this.getBeneficiaries(report.beneficiaries);
+
+        const messages = this.getSmsAndEmailMessages(report, donor, beneficiaries);
+
+        await Promise.all([
+          this.args.smsProvider.sendSms(donor.phone, messages.sms),
+          this.args.emailProvider.sendEmail(donor.email, messages.email),
+        ]);                 
+      });
+    }
+    catch(error) {
+      console.error('Error occurred when handling event', error);
+    }
+  }
+
+  private getSmsAndEmailMessages(report: DistributionReport, donor: User, beneficiaries: User[]): SmsAndEmailMessages {
+    const sms = `Hello ${donor.name},
```

Comment:
Let's just use the first name when addressing the user. I think there's utility function for extracting the firstName